functionary/prompt_template/__init__.py

```diff
@@ -37,16 +43,26 @@ def get_prompt_template_from_tokenizer(tokenizer: Any) -> PromptTemplate:
     p1 = PromptTemplateV1.get_prompt_template()
     p2 = PromptTemplateV2.get_prompt_template()
     p3 = Llama3Template.get_prompt_template()
+    p6 = Llama3TemplateV3.get_prompt_template()
+
     token_ids = tokenizer.encode(p3.function_separator, add_special_tokens=False)
+    if len(token_ids) == 1 and token_ids[0] == 128254:  # based on llam3
+        if p3.function_separator in tokenizer.chat_template:
+            return p3
+        else:
+            return p6
+
+    token_ids = tokenizer.encode(p2.from_token, add_special_tokens=False)
     if len(token_ids) == 1:
-        return p3
+        return p2
 
     token_ids = tokenizer.encode(p1.start_function, add_special_tokens=False)
     if token_ids[0] in [29871, 28705]:
         token_ids = token_ids[1:]
     if len(token_ids) == 1:
         return p1
-    return p2
+
+    raise Exception("Cannot detect prompt template based on tokenizer")
```

Comment:
Can we move this checking to each class as an abstract method?
```python
all_templates = [template.get_prompt_template() for template in PromptTemplate.__subclasses__()]
for template in all_templates:
    if template.confirm_template_given_tokenizer(tokenizer):
        return template
else:
    raise Exception("Cannot detect prompt template based on tokenizer")
```

Of course, feel free to change the name of the abstract method to your liking.

Reply:
Actually, I also think about adding a function for checking the tokenizer, but it would be hard some cases. For example, we have 2 templates without adding new tokens, template 1 uses: >> as function separator and template 2 uses: >>> as function separator.
In checking we will find: >> in chat_template and >>> in chat_template.
In checking, we should check template 2 first, then template 1.
If we check in a for-loop we are not sure about the order